bundle/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/TcsController.scala

```diff
@@ -0,0 +1,203 @@
+  case class NodChop(nod: Beam, chop: Beam)
+
+  // TCS can be configured to update a guide probe position only for certain nod-chop positions.
+  sealed trait NodChopTrackingConfig {
+    def get(nodchop: NodChop): Boolean
+  }
+
+  // If x is of type ActiveNodChopTracking then ∃ a:NodChop ∍ x.get(a)
+  // How could I reflect that in the code?
+  sealed trait ActiveNodChopTracking extends NodChopTrackingConfig
+
+  object NodChopTrackingConfig {
+
+    object None extends NodChopTrackingConfig {
+      def get(nodchop: NodChop): Boolean = false
+    }
+
+    object Normal extends ActiveNodChopTracking {
+      def get(nodchop: NodChop): Boolean = nodchop.nod == nodchop.chop
+    }
+
+    // Must be a non-empty set
+    case class Special(s: Set[NodChop]) extends ActiveNodChopTracking {
+      def get(nodchop: NodChop): Boolean = s(nodchop)
+    }
+
+  }
+
+  // A probe tracking configuration is specified by its nod-chop tracking table
+  // and its follow flag. The first tells TCS when to update the target track
+  // followed by the probe, and the second tells the probe when to move
+  // following the target track.
+  sealed trait ProbeTrackingConfig {
+    def follow: Boolean
+
+    def getNodChop: NodChopTrackingConfig
+  }
+
+  object ProbeTrackingConfig {
+
+    object Parked extends ProbeTrackingConfig {
+      override def follow: Boolean = false
+
+      override def getNodChop: NodChopTrackingConfig = NodChopTrackingConfig.None
+    }
+
+    object Off extends ProbeTrackingConfig {
+      override def follow: Boolean = false
+
+      override def getNodChop: NodChopTrackingConfig = NodChopTrackingConfig.None
+    }
+
+    case class On(ndconfig: ActiveNodChopTracking) extends ProbeTrackingConfig {
+      override def follow: Boolean = true
+
+      override def getNodChop: NodChopTrackingConfig = ndconfig
+    }
+
+  }
+
+  sealed trait HrwfsPickupPosition
+
+  object HrwfsPickupPosition {
+
+    object IN extends HrwfsPickupPosition
+
+    object OUT extends HrwfsPickupPosition
+
+    object Parked extends HrwfsPickupPosition
+
+  }
+
+  sealed trait LightSource
+
+  object LightSource {
+
+    object Sky extends LightSource
+
+    object AO extends LightSource
+
+    object GCAL extends LightSource
+
+  }
+
+  sealed trait ScienceFoldPosition
+
+  object ScienceFoldPosition {
+
+    import LightSource._
+
+    object Parked extends ScienceFoldPosition
+
+    class Position(source: LightSource, sink: Instrument) extends ScienceFoldPosition {
+      val sfPositionName: String = source match {
+        case Sky => sink.sfName
+        case AO => "ao2" + sink.sfName
+        case GCAL => "gcal2" + sink.sfName
+      }
+    }
+
+  }
+
+  case class GuideConfig(mountGuide: Boolean, m1Guide: M1GuideConfig, m2Guide: M2GuideConfig)
+
+  case class TelescopeConfig(offsetA: Offset, offsetB: Offset, offsetC: Offset, wavelA: Wavelength, wavelB: Wavelength,
```

Comment:
It's may be worth having distinct types for offsets and wavelengths in A/B/C, which will prevent them from getting mixed up. Similar with `GuidersTrackingConfig` below.